src/query/storage/m3/consolidators/filter.go

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package consolidators
+
+import (
+	"bytes"
+
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/x/ident"
+)
+
+func filterTagIterator(iter ident.TagIterator, filters models.Filters) bool {
+	if len(filters) == 0 || iter.Remaining() == 0 {
+		return false
+	}
+
+	// NB: rewind iterator for re-use.
+	defer iter.Rewind()
+	for iter.Next() {
+		tag := iter.Current()
+
+		name := tag.Name.Bytes()
+		value := tag.Value.Bytes()
+		for _, f := range filters {
+			if !bytes.Equal(name, f.Name) {
+				continue
+			}
+
+			// 0 length filters implies filtering for entire range.
+			if len(f.Values) == 0 {
+				return true
+			}
+
+			for _, filterValue := range f.Values {
+				if bytes.Equal(filterValue, value) {
+					return true
+				}
+			}
+		}
+	}
+
+	return false
+}
+
+func filterNames(tags []CompletedTag, filters models.Filters) []CompletedTag {
+	if len(filters) == 0 || len(tags) == 0 {
+		return tags
+	}
+
+	filteredTags := tags[:0]
+	skip := false
+	for _, tag := range tags {
+		skip = false
+		for _, f := range filters {
+			if len(f.Values) != 0 {
+				// If this has filter values, it is not a name filter, and the result
+				// is valid.
```

Comment:
Hm wouldn't we want to skip this tag name though if tag.Name matches f.Name?  i.e. if we don't want to see any time series with "foo_k8s_namespace" then we don't want the "foo_k8s_namespace" tag name to come back either right?